x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java

```diff
@@ -391,11 +392,15 @@ else if (child instanceof GroupingFunction) {
             if (f instanceof Count) {
                 Count c = (Count) f;
                 // COUNT(*) or COUNT(<literal>)
-                if (c.field() instanceof Literal) {
+                if (c.field().foldable()) {
                     AggRef ref = groupingAgg == null ?
                             GlobalCountRef.INSTANCE :
                             new GroupByRef(groupingAgg.id(), Property.COUNT, null);
 
+                    // if the count points to the total track hits, enable accurate count retrieval
+                    if (ref == GlobalCountRef.INSTANCE) {
```

Comment:
I would add this to an if block in combination with the above check on `groupingAgg == null`.